src/serial.rs

```diff
@@ -0,0 +1,484 @@
+            .pop()
+            .expect("called serial::write with no expectation")
+        {
+            Mode::Write(expectation) => {
+                assert_eq!(
+                    expectation, word,
+                    "serial::write expected to write {:?} but actually wrote {:?}",
+                    expectation, word
+                );
+                Ok(())
+            }
+            mode => panic!(
+                "expected to perform a serial transaction '{:?}' but instead did a write of {:?}",
+                mode, word
+            ),
+        }
+    }
+
+    fn flush(&mut self) -> nb::Result<(), Self::Error> {
+        match self
+            .pop()
+            .expect("called serial::flush with no expectation")
+        {
+            Mode::Flush => Ok(()),
+            mode => panic!(
+                "expected to perform a serial transaction '{:?}' but instead did a flush",
+                mode
+            ),
+        }
+    }
+}
+
+// Note: we attempted to provide our own implementation of
+// embedded_hal::blocking::serial::Write. However, we're unable
+// to override it due to the blanket default implementation provided by
+// the embedded_hal crate. It comes down to the fact that, if we were
+// to provide an embedded_hal::blocking::serial::Write implementation
+// here, any user of embedded_hal would be free to implement the *defuault*
+// version for our type. Therefore, we conform to the default implementation,
+// knowing that the default is implemented in terms of the non-blocking
+// trait, which is defined above.
+//
+// If you know a way around this, please let us know!
+impl<'a, Word> write::Default<Word> for Mock<'a, Word> where
+    Word: PartialEq + std::fmt::Debug + Clone
+{
+}
+
+#[cfg(test)]
+mod test {
+    use super::Mock;
+    use super::Transaction;
+    use embedded_hal::blocking::serial::Write as BWrite;
+    use embedded_hal::serial::{Read, Write};
+
+    #[test]
+    fn test_serial_mock_read() {
+        let ts = [Transaction::read(0x54)];
+        let mut ser = Mock::new(&ts);
+        let r = ser.read().expect("failed to read");
+        assert_eq!(r, 0x54);
+        ser.done();
+    }
+
+    #[test]
+    fn test_serial_mock_write_single_value_nonblocking() {
+        let ts = [Transaction::write([0xAB])];
+        let mut ser = Mock::new(&ts);
+        ser.write(0xAB).unwrap();
+        ser.done();
+    }
+
+    #[test]
+    fn test_serial_mock_write_many_values_nonblocking() {
+        let ts = [Transaction::write([0xAB, 0xCD, 0xEF])];
+        let mut ser = Mock::new(&ts);
+        ser.write(0xAB).unwrap();
+        ser.write(0xCD).unwrap();
+        ser.write(0xEF).unwrap();
+        ser.done();
+    }
+
+    #[test]
+    fn test_serial_mock_blocking_write() {
+        let ts = [Transaction::write([0xAB, 0xCD, 0xEF])];
+        let mut ser = Mock::new(&ts);
+        ser.bwrite_all(&[0xAB, 0xCD, 0xEF]).unwrap();
+        ser.done();
+    }
+
+    #[test]
+    #[should_panic]
```

Comment:
You should be able to assert a certain panic message. Example:

> #[should_panic(expected = "Divide result is zero")]

See https://doc.rust-lang.org/rust-by-example/testing/unit_testing.html#testing-panics

Reply:
I'll add them.

Are we looking to guarantee the panic messages as part of the public API? Otherwise, what value will this provide? This might seem inconsistent with the SPI and I2C mock testing approaches.

Reply:
Yep, exactly :slightly_smiling_face: This is not about guaranteeing panic messages, it's about testing that the right panics happen.